go/disk.go

```diff
@@ -0,0 +1,352 @@
+	res := fmt.Sprintf("%s,%s", diskDriver(d.Trim), d.Path)
+	if d.Format != "" {
+		res += ",format=" + d.Format
+	}
+	return res
+}
+
+/*-----------.
+| QcowDisk.  |
+`-----------*/
+
+// QcowDisk describes a qcow2 disk image file.
+type QcowDisk struct {
+	// Path specifies where the image file will be.
+	Path string `json:"path"`
+	// Size specifies the size of the disk image.  Used if the image needs to be created.
+	Size int `json:"size"`
+	// Format is passed as-is to the driver.
+	Format string `json:"format"`
+	// Trim specifies whether we should trim the image file.
+	Trim bool `json:"trim"`
+	// QcowToolPath is the path to the binary to use to manage this image.
+	// Defaults to "qcow-tool" when empty.
+	QcowToolPath   string
+	OnFlush        string
+	CompactAfter   int
+	KeepErased     int
+	RuntimeAsserts bool
+	Stats          string
+}
+
+// GetPath returns the location of the disk image file.
+func (d *QcowDisk) GetPath() string {
+	return d.Path
+}
+
+// SetPath changes the location of the disk image file.
+func (d *QcowDisk) SetPath(p string) {
+	d.Path = p
+}
+
+// GetSize returns the desired size of the disk image file.
+func (d *QcowDisk) GetSize() int {
+	return d.Size
+}
+
+// String returns the path.
+func (d *QcowDisk) String() string {
+	return d.Path
+}
+
+// QcowTool prepares a call to qcow-tool on this image.
+func (d *QcowDisk) QcowTool(verb string, args ...string) *exec.Cmd {
+	path := d.QcowToolPath
+	if path == "" {
+		path = "qcow-tool"
+	}
+	return exec.Command(path, append([]string{verb, d.Path}, args...)...)
+}
+
+// Exists if the image file exists.
+func (d *QcowDisk) Exists() bool {
+	return exists(d)
+}
+
+// Ensure creates the disk image if needed, and resizes it if needed.
+func (d *QcowDisk) Ensure() error {
+	return ensure(d)
+}
+
+// Create a disk with the given size in MiB
+func (d *QcowDisk) create() error {
+	cmd := d.QcowTool("create", "--size", fmt.Sprintf("%dMiB", d.Size))
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	return cmd.Wait()
+}
+
+// Query the current virtual size of the disk in MiB
+func (d *QcowDisk) getFileSize() (int, error) {
+	// Return a failure if the file doesn't exist yet
+	if _, err := os.Stat(d.Path); err != nil {
+		return 0, err
+	}
+	cmd := d.QcowTool("info", "--filter", ".size")
+	out, err := cmd.Output()
+	size, err := strconv.ParseInt(strings.TrimSpace(string(out)), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+	return int(size / mib), nil
+}
+
+// Resize the virtual size of the disk
+func (d *QcowDisk) resize() error {
+	cmd := d.QcowTool("resize", "--size", fmt.Sprintf("%dMiB", d.Size))
+	return cmd.Run()
+}
+
+func (d *QcowDisk) sizeString() string {
+	s, err := d.getFileSize()
+	if err != nil {
+		return fmt.Sprintf("cannot get size: %v", err)
+	}
+	return fmt.Sprintf("%vMiB", s)
+}
+
+// compact the disk to shrink the physical size.
+func (d *QcowDisk) compact(lockFh *os.File) error {
+	log.Infof("Compact: %q... (%v)", d, d.sizeString())
```

Comment:
You format strings with `%v` quite a lot in this PR, is there any reason to prefer `%v` over `%s`? If not (and I don't think there is, but perhaps I'm missing something) please use `%s`.

Reply:
Well, this is Go, not C.  We have decent variadic functions that do not require the users to do the compiler's job of tracking types.  Boost.Format, Lisp, Python, Rust, etc. all go away from the %d/%s model.  Go's universal format spec is %v.

What's to be preferred with %s?

Reply:
Mainly that `%s` doesn't need to do a bunch of runtime introspection, probably doesn't matter for this case, lets leave it then.

Reply:
Good point!  But since it's all based on `interface{}`, it must do that anyway I guess.  It doesn't seem to make much of a difference.
```go
package foo

import (
	"fmt"
	"testing"
)

func BenchmarkVString(b *testing.B) {
	s := "Hello, world!"
	for i := 0; i < b.N; i++ {
		fmt.Sprintf("%v", s)
	}
}

func BenchmarkVInt(b *testing.B) {
	s := 42
	for i := 0; i < b.N; i++ {
		fmt.Sprintf("%v", s)
	}
}

func BenchmarkSString(b *testing.B) {
	s := "Hello, world!"
	for i := 0; i < b.N; i++ {
		fmt.Sprintf("%s", s)
	}
}

func BenchmarkDInt(b *testing.B) {
	s := 42
	for i := 0; i < b.N; i++ {
		fmt.Sprintf("%d", s)
	}
}
```
```
go test -bench .
goos: darwin
goarch: amd64
BenchmarkVString-8   	10000000	       109 ns/op
BenchmarkVInt-8      	20000000	        97.8 ns/op
BenchmarkSString-8   	10000000	       112 ns/op
BenchmarkDInt-8      	20000000	        97.7 ns/op
```